GitHub CI is running out of space

GitHub CI is consistently failing with errors such as:

```
 [8 of 9] Compiling Format           ( test/functional/Format.hs, /home/runner/work/haskell-language-server/haskell-language-server/dist-newstyle/build/x86_64-linux/ghc-9.4.8/hls-2.9.0.1/t/func-test/build/func-test/func-test-tmp/Format.o )
[1 of 1] Compiling Main             ( test/wrapper/Main.hs, /home/runner/work/haskell-language-server/haskell-language-server/dist-newstyle/build/x86_64-linux/ghc-9.4.8/hls-2.9.0.1/t/wrapper-test/build/wrapper-test/wrapper-test-tmp/Main.o )
[9 of 9] Compiling Main             ( test/functional/Main.hs, /home/runner/work/haskell-language-server/haskell-language-server/dist-newstyle/build/x86_64-linux/ghc-9.4.8/hls-2.9.0.1/t/func-test/build/func-test/func-test-tmp/Main.o )
[2 of 2] Linking /home/runner/work/haskell-language-server/haskell-language-server/dist-newstyle/build/x86_64-linux/ghc-9.4.8/hls-2.9.0.1/b/benchmark/build/benchmark/benchmark

<no location info>: warning: [-Wunused-packages]
    The following packages were specified via -package or -package-id flags,
    but were not needed for compilation:
      - hls-2.9.0.1 (exposed by flag -package-id hls-2.9.0.1-inplace-hls-refactor-plugin)
[2 of 2] Linking /home/runner/work/haskell-language-server/haskell-language-server/dist-newstyle/build/x86_64-linux/ghc-9.4.8/hls-2.9.0.1/t/wrapper-test/build/wrapper-test/wrapper-test
[10 of 10] Linking /home/runner/work/haskell-language-server/haskell-language-server/dist-newstyle/build/x86_64-linux/ghc-9.4.8/hls-2.9.0.1/t/func-test/build/func-test/func-test
/usr/bin/ld.gold: fatal error: /home/runner/work/haskell-language-server/haskell-language-server/dist-newstyle/build/x86_64-linux/ghc-9.4.8/hls-2.9.0.1/b/benchmark/build/benchmark/benchmark: No space left on device
collect2: error: ld returned 1 exit status
ghc-9.4.8: `gcc' failed in phase `Linker'. (Exit code: 1)
Error: cabal: Failed to build bench:benchmark from hls-2.9.0.1.
```

See https://github.com/haskell/haskell-language-server/actions/runs/10446351088/job/28923644007

--- a/.github/actions/setup-build/action.yml
+++ b/.github/actions/setup-build/action.yml
@@ -116,3 +116,18 @@ runs:
     - name: "Remove freeze file"
       run: rm -f cabal.project.freeze
       shell: bash
+
+    # Make sure to clear all unneeded `ghcup`` caches.
+    # At some point, we were running out of disk space, see issue
+    # https://github.com/haskell/haskell-language-server/issues/4386 for details.
+    #
+    # Using "printf" debugging (`du -sh *` and `df -h /`) and binary searching,
+    # we figured out that `ghcup` caches are taking up a sizable portion of the
+    # disk space.
+    # Thus, we remove anything we don't need, especially caches and temporary files.
+    # For got measure, we also make sure no other tooling versions are
+    # installed besides the ones we explicitly want.
+    - name: "Remove ghcup caches"
+      if: runner.os == 'Linux'
+      run: ghcup gc --ghc-old --share-dir --hls-no-ghc --cache --tmpdirs --unset
+      shell: bash

